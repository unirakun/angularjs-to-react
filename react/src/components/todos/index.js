export { default } from './todos.container'
