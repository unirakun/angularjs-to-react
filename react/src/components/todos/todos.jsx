import React from 'react'
import PropTypes from 'prop-types'
import { onlyUpdateForPropTypes } from 'recompose'
import Todo from 'components/todo'

const Todos = ({ todos, allCompleted, onCompleteAll }) => {
  return (
    <section className="main">
      {todos.length > 0 &&
        <input
          className="toggle-all"
          type="checkbox"
          onChange={onCompleteAll}
          checked={allCompleted}
        />
      }
      <ul className="todo-list" >
        {todos.map(todo => <Todo key={todo} id={todo} />)}
      </ul>
    </section>
  )
}

Todos.propTypes = {
  todos: PropTypes.array,
  allCompleted: PropTypes.bool.isRequired,
  onCompleteAll: PropTypes.func.isRequired,
}

Todos.defaultProps = {
  todos: [],
}

export default onlyUpdateForPropTypes(Todos)
