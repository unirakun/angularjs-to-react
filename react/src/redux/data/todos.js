import { keyValue } from 'k-redux-factory'
import { mapPayload } from 'k-redux-factory/helpers'

/* Todos Reducer  with her mapper */
export const todosMapper = todo => ({ id: new Date().getTime(), todo: todo.trim() })
export default keyValue({ pre: [mapPayload(/@@krf\/ADD_TODOS/)(todosMapper)] })({ path: 'data', key: 'id', name: 'todos' })
