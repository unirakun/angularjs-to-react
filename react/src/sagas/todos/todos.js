import { call, put, select } from 'redux-saga/effects'
import store from 'redux/data/todos'
import completed from 'redux/ui/completed'
import editing from 'redux/ui/editing'

export function goToAngular() {
  window.location = '/angularjs'
}

export function* remove(todoId) {
  yield put(store.remove(todoId))
  yield put(editing.remove(todoId))
  yield put(completed.remove(todoId))
}

export function* update(todoId) {
  const todo = yield select(editing.get(todoId))
  if (!todo) return
  if (!todo.todo || !todo.todo.trim()) yield call(remove, todoId)
  else yield put(store.update(todo))
  yield put(editing.remove(todoId))
}

export function* complete(todo) {
  yield put(completed.add(todo))
}

export function* completeAll(check) {
  const todos = yield select(store.getAsArray)
  yield* todos.map(todo => call(complete, { ...todo, completed: check }))
}

export function* clearCompleted() {
  const completeTodos = yield select(completed.getBy('completed', true))
  const completedId = completeTodos.map(todo => todo.id)
  yield call(remove, completedId)
}
