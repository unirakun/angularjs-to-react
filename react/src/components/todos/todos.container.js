import { connect } from 'react-redux'
import todos from 'redux/data/todos'
import completed from 'redux/ui/completed'
import action from 'components/actionFactory'
import Component from './todos'

export const mapStateToProps = (state) => {
  return {
    todos: todos.getKeys(state),
    allCompleted: completed.getBy('completed', true)(state).length === todos.getLength(state),
  }
}

export const mapDispatchToProps = (dispatch) => {
  return {
    onCompleteAll: e => dispatch(action('COMPLETE_ALL_CHANGED', e.target.checked)),
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Component)
