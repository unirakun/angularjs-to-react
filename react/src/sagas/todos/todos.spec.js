/* eslint-env jest */
import tester from 'k-redux-saga-tester'
import { add, remove, update, complete, completeAll, clearCompleted } from './todos'

describe('sagas/todos', () => {
  describe('add', () => {
    const test = tester(add)

    it('should not add todo when todo is null', () => {
      expect(test(null)()).toMatchSnapshot()
    })

    it('should not add todo when todo is undefined', () => {
      expect(test(undefined)()).toMatchSnapshot()
    })

    it('should not add todo when todo is empty', () => {
      expect(test('')()).toMatchSnapshot()
    })

    it('should not add todo when todo is blank', () => {
      expect(test('  ')()).toMatchSnapshot()
    })

    it('should add todo when todo is valid', () => {
      expect(test('todo1')()).toMatchSnapshot()
    })
  })

  describe('remove', () => {
    const test = tester(remove)

    it('should remove todo', () => {
      expect(test('todo1')()).toMatchSnapshot()
    })
  })

  describe('update', () => {
    const test = tester(update)

    it('should not update todo when todo does not exist', () => {
      const mocks = {
        select: [() => undefined],
        call: [() => false],
      }
      expect(test('todoDoesNotExist')(mocks)).toMatchSnapshot()
    })

    it('should remove todo when the todo is empty', () => {
      const mocks = {
        select: [() => ({ id: 'todo1', todo: '' })],
        call: [() => false],
      }
      expect(test('todo1')(mocks)).toMatchSnapshot()
    })

    it('should remove todo when the todo is blank', () => {
      const mocks = {
        select: [() => ({ id: 'todo1', todo: '  ' })],
        call: [() => false],
      }
      expect(test('todo1')(mocks)).toMatchSnapshot()
    })

    it('should update todo', () => {
      const mocks = {
        select: [() => ({ id: 'todo1', todo: 'todoOnEdit' })],
        call: [() => false],
      }
      expect(test('todo1')(mocks)).toMatchSnapshot()
    })
  })

  describe('complete', () => {
    const test = tester(complete)

    it('should complete todo', () => {
      expect(test('todo1')()).toMatchSnapshot()
    })
  })

  describe('completeAll', () => {
    const test = tester(completeAll)

    it('should complete one todos', () => {
      const mocks = {
        select: [() => ([{ id: 'todo1', todo: 'todoOnEdit' }])],
        call: [() => true],
      }
      expect(test(true)(mocks)).toMatchSnapshot()
      expect(test(false)(mocks)).toMatchSnapshot()
    })

    it('should complete many todos', () => {
      const mocks = {
        select: [() => [
          { id: 'todo1', todo: 'todo1' },
          { id: 'todo2', todo: 'todo2' },
        ]],
        call: [() => 'todo1', () => 'todo2'],
      }
      expect(test(true)(mocks)).toMatchSnapshot()
      expect(test(false)(mocks)).toMatchSnapshot()
    })
  })

  describe('clearCompleted', () => {
    const test = tester(clearCompleted)

    it('should clear completed todo', () => {
      const mocks = {
        select: [() => [
          { id: 'todo1', todo: 'todo1' },
          { id: 'todo2', todo: 'todo2' },
        ]],
        call: [() => false],
      }

      expect(test('todo1')(mocks)).toMatchSnapshot()
    })
  })
})
