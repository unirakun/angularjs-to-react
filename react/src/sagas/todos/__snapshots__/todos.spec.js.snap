// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sagas/todos add should add todo when todo is valid 1`] = `
Array [
  Object {
    "done": false,
    "value": Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": "todo1",
          "type": "@@krf/ADD_TODOS",
        },
        "channel": null,
      },
    },
  },
  Object {
    "done": false,
    "value": Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "type": "@@krf/RESET_NEWTODO",
        },
        "channel": null,
      },
    },
  },
  Object {
    "done": true,
    "value": undefined,
  },
]
`;

exports[`sagas/todos add should not add todo when todo is blank 1`] = `
Array [
  Object {
    "done": true,
    "value": undefined,
  },
]
`;

exports[`sagas/todos add should not add todo when todo is empty 1`] = `
Array [
  Object {
    "done": true,
    "value": undefined,
  },
]
`;

exports[`sagas/todos add should not add todo when todo is null 1`] = `
Array [
  Object {
    "done": true,
    "value": undefined,
  },
]
`;

exports[`sagas/todos add should not add todo when todo is undefined 1`] = `
Array [
  Object {
    "done": true,
    "value": undefined,
  },
]
`;

exports[`sagas/todos clearCompleted should clear completed todo 1`] = `
Array [
  Object {
    "raw": Object {
      "done": false,
      "value": Object {
        "@@redux-saga/IO": true,
        "SELECT": Object {
          "args": Array [],
          "selector": [Function],
        },
      },
    },
    "value": Array [
      Object {
        "id": "todo1",
        "todo": "todo1",
      },
      Object {
        "id": "todo2",
        "todo": "todo2",
      },
    ],
  },
  Object {
    "raw": Object {
      "done": false,
      "value": Object {
        "@@redux-saga/IO": true,
        "CALL": Object {
          "args": Array [
            Array [
              "todo1",
              "todo2",
            ],
          ],
          "context": null,
          "fn": [Function],
        },
      },
    },
    "value": false,
  },
  Object {
    "done": true,
    "value": undefined,
  },
]
`;

exports[`sagas/todos complete should complete todo 1`] = `
Array [
  Object {
    "done": false,
    "value": Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": "todo1",
          "type": "@@krf/ADD_COMPLETED",
        },
        "channel": null,
      },
    },
  },
  Object {
    "done": true,
    "value": undefined,
  },
]
`;

exports[`sagas/todos completeAll should complete many todos 1`] = `
Array [
  Object {
    "raw": Object {
      "done": false,
      "value": Object {
        "@@redux-saga/IO": true,
        "SELECT": Object {
          "args": Array [],
          "selector": [Function],
        },
      },
    },
    "value": Array [
      Object {
        "id": "todo1",
        "todo": "todo1",
      },
      Object {
        "id": "todo2",
        "todo": "todo2",
      },
    ],
  },
  Object {
    "raw": Object {
      "done": false,
      "value": Object {
        "@@redux-saga/IO": true,
        "CALL": Object {
          "args": Array [
            Object {
              "completed": true,
              "id": "todo1",
              "todo": "todo1",
            },
          ],
          "context": null,
          "fn": [Function],
        },
      },
    },
    "value": "todo1",
  },
  Object {
    "raw": Object {
      "done": false,
      "value": Object {
        "@@redux-saga/IO": true,
        "CALL": Object {
          "args": Array [
            Object {
              "completed": true,
              "id": "todo2",
              "todo": "todo2",
            },
          ],
          "context": null,
          "fn": [Function],
        },
      },
    },
    "value": "todo2",
  },
  Object {
    "done": true,
    "value": undefined,
  },
]
`;

exports[`sagas/todos completeAll should complete many todos 2`] = `
Array [
  Object {
    "raw": Object {
      "done": false,
      "value": Object {
        "@@redux-saga/IO": true,
        "SELECT": Object {
          "args": Array [],
          "selector": [Function],
        },
      },
    },
    "value": Array [
      Object {
        "id": "todo1",
        "todo": "todo1",
      },
      Object {
        "id": "todo2",
        "todo": "todo2",
      },
    ],
  },
  Object {
    "raw": Object {
      "done": false,
      "value": Object {
        "@@redux-saga/IO": true,
        "CALL": Object {
          "args": Array [
            Object {
              "completed": false,
              "id": "todo1",
              "todo": "todo1",
            },
          ],
          "context": null,
          "fn": [Function],
        },
      },
    },
    "value": "todo1",
  },
  Object {
    "raw": Object {
      "done": false,
      "value": Object {
        "@@redux-saga/IO": true,
        "CALL": Object {
          "args": Array [
            Object {
              "completed": false,
              "id": "todo2",
              "todo": "todo2",
            },
          ],
          "context": null,
          "fn": [Function],
        },
      },
    },
    "value": "todo2",
  },
  Object {
    "done": true,
    "value": undefined,
  },
]
`;

exports[`sagas/todos completeAll should complete one todos 1`] = `
Array [
  Object {
    "raw": Object {
      "done": false,
      "value": Object {
        "@@redux-saga/IO": true,
        "SELECT": Object {
          "args": Array [],
          "selector": [Function],
        },
      },
    },
    "value": Array [
      Object {
        "id": "todo1",
        "todo": "todoOnEdit",
      },
    ],
  },
  Object {
    "raw": Object {
      "done": false,
      "value": Object {
        "@@redux-saga/IO": true,
        "CALL": Object {
          "args": Array [
            Object {
              "completed": true,
              "id": "todo1",
              "todo": "todoOnEdit",
            },
          ],
          "context": null,
          "fn": [Function],
        },
      },
    },
    "value": true,
  },
  Object {
    "done": true,
    "value": undefined,
  },
]
`;

exports[`sagas/todos completeAll should complete one todos 2`] = `
Array [
  Object {
    "raw": Object {
      "done": false,
      "value": Object {
        "@@redux-saga/IO": true,
        "SELECT": Object {
          "args": Array [],
          "selector": [Function],
        },
      },
    },
    "value": Array [
      Object {
        "id": "todo1",
        "todo": "todoOnEdit",
      },
    ],
  },
  Object {
    "raw": Object {
      "done": false,
      "value": Object {
        "@@redux-saga/IO": true,
        "CALL": Object {
          "args": Array [
            Object {
              "completed": false,
              "id": "todo1",
              "todo": "todoOnEdit",
            },
          ],
          "context": null,
          "fn": [Function],
        },
      },
    },
    "value": true,
  },
  Object {
    "done": true,
    "value": undefined,
  },
]
`;

exports[`sagas/todos remove should remove todo 1`] = `
Array [
  Object {
    "done": false,
    "value": Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": "todo1",
          "type": "@@krf/REMOVE_TODOS",
        },
        "channel": null,
      },
    },
  },
  Object {
    "done": false,
    "value": Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": "todo1",
          "type": "@@krf/REMOVE_EDITING",
        },
        "channel": null,
      },
    },
  },
  Object {
    "done": false,
    "value": Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": "todo1",
          "type": "@@krf/REMOVE_COMPLETED",
        },
        "channel": null,
      },
    },
  },
  Object {
    "done": true,
    "value": undefined,
  },
]
`;

exports[`sagas/todos update should not update todo when todo does not exist 1`] = `
Array [
  Object {
    "raw": Object {
      "done": false,
      "value": Object {
        "@@redux-saga/IO": true,
        "SELECT": Object {
          "args": Array [],
          "selector": [Function],
        },
      },
    },
    "value": undefined,
  },
  Object {
    "done": true,
    "value": undefined,
  },
]
`;

exports[`sagas/todos update should remove todo when the todo is blank 1`] = `
Array [
  Object {
    "raw": Object {
      "done": false,
      "value": Object {
        "@@redux-saga/IO": true,
        "SELECT": Object {
          "args": Array [],
          "selector": [Function],
        },
      },
    },
    "value": Object {
      "id": "todo1",
      "todo": "  ",
    },
  },
  Object {
    "raw": Object {
      "done": false,
      "value": Object {
        "@@redux-saga/IO": true,
        "CALL": Object {
          "args": Array [
            "todo1",
          ],
          "context": null,
          "fn": [Function],
        },
      },
    },
    "value": false,
  },
  Object {
    "done": false,
    "value": Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": "todo1",
          "type": "@@krf/REMOVE_EDITING",
        },
        "channel": null,
      },
    },
  },
  Object {
    "done": true,
    "value": undefined,
  },
]
`;

exports[`sagas/todos update should remove todo when the todo is empty 1`] = `
Array [
  Object {
    "raw": Object {
      "done": false,
      "value": Object {
        "@@redux-saga/IO": true,
        "SELECT": Object {
          "args": Array [],
          "selector": [Function],
        },
      },
    },
    "value": Object {
      "id": "todo1",
      "todo": "",
    },
  },
  Object {
    "raw": Object {
      "done": false,
      "value": Object {
        "@@redux-saga/IO": true,
        "CALL": Object {
          "args": Array [
            "todo1",
          ],
          "context": null,
          "fn": [Function],
        },
      },
    },
    "value": false,
  },
  Object {
    "done": false,
    "value": Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": "todo1",
          "type": "@@krf/REMOVE_EDITING",
        },
        "channel": null,
      },
    },
  },
  Object {
    "done": true,
    "value": undefined,
  },
]
`;

exports[`sagas/todos update should update todo 1`] = `
Array [
  Object {
    "raw": Object {
      "done": false,
      "value": Object {
        "@@redux-saga/IO": true,
        "SELECT": Object {
          "args": Array [],
          "selector": [Function],
        },
      },
    },
    "value": Object {
      "id": "todo1",
      "todo": "todoOnEdit",
    },
  },
  Object {
    "done": false,
    "value": Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": Object {
            "id": "todo1",
            "todo": "todoOnEdit",
          },
          "type": "@@krf/UPDATE_TODOS",
        },
        "channel": null,
      },
    },
  },
  Object {
    "done": false,
    "value": Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": "todo1",
          "type": "@@krf/REMOVE_EDITING",
        },
        "channel": null,
      },
    },
  },
  Object {
    "done": true,
    "value": undefined,
  },
]
`;
